groups/nts/ntsa/ntsa_interest.h

```diff
@@ -0,0 +1,1189 @@
+    /// optionally specified indentation 'level' and return a reference to
+    /// the modifiable 'stream'.  If 'level' is specified, optionally
+    /// specify 'spacesPerLevel', the number of spaces per indentation level
+    /// for this and all of its nested objects.  Each line is indented by
+    /// the absolute value of 'level * spacesPerLevel'.  If 'level' is
+    /// negative, suppress indentation of the first line.  If
+    /// 'spacesPerLevel' is negative, suppress line breaks and format the
+    /// entire output on one line.  If 'stream' is initially invalid, this
+    /// operation has no effect.  Note that a trailing newline is provided
+    /// in multiline mode only.
+    bsl::ostream& print(bsl::ostream& stream,
+                        int           level          = 0,
+                        int           spacesPerLevel = 4) const;
+
+    /// Defines the traits of this type. These traits can be used to
+    /// select, at compile-time, the most efficient algorithm to manipulate
+    /// objects of this type.
+    NTSCFG_DECLARE_NESTED_BITWISE_MOVABLE_TRAITS(Interest);
+};
+
+/// Insert a formatted, human-readable description of the specified 'value'
+/// into the specified 'stream'. Return a reference to the modifiable
+/// 'stream'.
+///
+/// @related ntsa::Interest
+bsl::ostream& operator<<(bsl::ostream& stream, const Interest& value);
+
+/// Return true if the specified 'lhs' has the same value as the specified
+/// 'rhs', otherwise return false.
+///
+/// @related ntsa::Interest
+bool operator==(const Interest& lhs, const Interest& rhs);
+
+/// Return true if the specified 'lhs' does not have the same value as the
+/// specified 'rhs', otherwise return false.
+///
+/// @related ntsa::Interest
+bool operator!=(const Interest& lhs, const Interest& rhs);
+
+/// Return true if the value of the specified 'lhs' is less than the value
+/// of the specified 'rhs', otherwise return false.
+///
+/// @related ntsa::Interest
+bool operator<(const Interest& lhs, const Interest& rhs);
+
+/// Contribute the values of the salient attributes of the specified 'value'
+/// to the specified hash 'algorithm'.
+///
+/// @related ntsa::Interest
+template <typename HASH_ALGORITHM>
+void hashAppend(HASH_ALGORITHM& algorithm, const Interest& value);
+
+/// Provide a set of interest.
+///
+/// @details
+/// This class provides a data structure to represent a set of interest in the
+/// state of a socket.
+///
+/// @par Thread Safety
+/// This class is not thread safe.
+///
+/// @ingroup module_ntsa_system
+class InterestSet
+{
+    class ConstIterator;
+    friend class ConstIterator;
+
+    class Iterator;
+    friend class Iterator;
+
+    /// Define a type alias for of vector of interest, indexed by socket
+    /// handle.
+    typedef bsl::vector<ntsa::Interest> Vector;
+
+    /// Define a type alias for a set of socket handles.
+    typedef bsl::set<ntsa::Handle> Set;
+
+    Vector            d_vector;
+    Set               d_set;
```

Comment:
As far as I got it `d_set` is used to only prevent a socket to be attached twice, to control detachment process and to count numer of attached sockets. Probably, the same thing can be achieved if type of `Vector` is changed to something like `typedef bsl::vector<bsl::pair<bool,. ntsa::Interest> > Vector;` where the first member of the pair means status of the socket (attached/detached) and a separate socket counter.

Reply:
ntsa::Interest and ntsa::Event and their respective sets are designed to store and merge what we want (ntsa::Interest) with what we have (ntsa::Event) with some guarantees of order of traversal. I want to position these components to be usable in both test drivers, where clarity of operating system behavior is paramount, as well as aiding us in revising our implementation of dynamic load balancing in NTC, where performance is critical. Right now, the interest and event sets are implemented very simply, to provide the basic design. Next up is a revision of their implementations into something like a giant bit set, leveraging the fact that file descriptor numbers are always kept fairly small, while still providing us O(1) lookups and guaranteed order of traversal (and perhaps queue-like behavior ala epoll and kqueue, but that may be getting ahead of ourselves.)